### PR TITLE
Only make the query to get all tables once

### PR DESCRIPTION
**Pull request by markstuart**
On larger datasets, the query was timing out
or taking a very long time, and making that
query every 100 records was taking forever to get
through the script. This will still fail with too
many records, as the _table_metadata view does not
do a true limit, it has to get all the data first
and then give back the first n rows. So the query
takes ages regardless of the limit.

This work improves the readability of the script as well (I hope).

### CODE REVIEW

File: ckanext/dia/commands.py [hunk 1]
```diff
@@ -69,22 +57,55 @@ def cleanup_datastore_once(self, start_offset):
...
+        print('Total errors when attempting deletion: {}'.format(len(failed_deletes)))
+        if len(failed_deletes) > 0 and len(failed_deletes) < 10: # don't print if too many
+            print('Datastore table names that failed to delete: {}'.format(failed_deletes))
+
+    def _get_all_datastore_tables(self, context):
+        result_set = logic.get_action('datastore_search')(
+            context,
+            {
+                'resource_id': '_table_metadata',
+                'limit': 50000, # arbitrarily high, but default limit is 100
```
Comment (markstuart) [inline ckanext/dia/commands.py:100]: I suspect the query would time out before 50000 rows anyway.

File: ckanext/dia/commands.py [hunk 2]
```diff
@@ -69,22 +57,55 @@ def cleanup_datastore_once(self, start_offset):
...
+            time.sleep(1)
+
+        updated_table_count = len(self._get_all_datastore_tables(context))
+
+        print('Cleanup complete!')
+        print('Total datastore tables before: {} and after: {}'.format(len(all_datastore_tables), updated_table_count))
+        print('Total orphaned datastore tables deleted: {}'.format(total_deleted))
+        print('Total errors when attempting deletion: {}'.format(len(failed_deletes)))
+        if len(failed_deletes) > 0 and len(failed_deletes) < 10: # don't print if too many
+            print('Datastore table names that failed to delete: {}'.format(failed_deletes))
```
Comment (markstuart) [inline ckanext/dia/commands.py:93]: Improved reporting so that we get totals at the end of the run, reduces some noise.

File: ckanext/dia/commands.py [hunk 3]
```diff
@@ -69,22 +57,55 @@ def cleanup_datastore_once(self, start_offset):
...
+                failed_deletes.extend(error_list)
+                print('Deleted in this chunk: {}'.format(delete_count))
+                print('Errored in this chunk: {}'.format(len(error_list)))
+
+            if end_chunk_index >= len(all_datastore_tables):
                 break
 
-        print('Total resources missing but referenced in datastore: {}'.format(len(resource_id_list)))
+            print('{} tables checked, {} more to go...'.format(end_chunk_index, len(all_datastore_tables) - end_chunk_index))
+            time.sleep(1)
```
Comment (markstuart) [inline ckanext/dia/commands.py:84]: In the cases where there aren't any deletes to be made, the db connection seems to lock up if we just loop over all the items testing whether the resource exists, so giving it a little break here between chunks.

File: ckanext/dia/commands.py [hunk 4]
```diff
@@ -69,22 +57,55 @@ def cleanup_datastore_once(self, start_offset):
...
-        for offset in itertools.count(start=start_offset, step=100):
-            print "Load metadata records from datastore (offset: %s)" % offset
-            record_list, tested = self._get_datastore_table_page(context, offset)  # noqa
-            total_tested += tested
-            resource_id_list.extend(record_list)
-            if not tested > 0:
-                break
-            if len(resource_id_list) > 250:
+        print("Loading metadata records from datastore")
+        all_datastore_tables = self._get_all_datastore_tables(context)
```
Comment (markstuart) [inline ckanext/dia/commands.py:61]: Get all the records in one call, then iterate over chunks below.
Each datastore table name in a chunk is then tested to see if there is a corresponding resource in the resource table, and it is marked as an orphan if none is found.
After building a list of orphaned datastore tables for the current chunk, they are deleted and count of success/failures tallied.
When all chunks are done, report.

Comment (markstuart) [reply ckanext/dia/commands.py:61]: The actual intent of the script hasn't really changed from the original, just the way that the dataset is retrieved and iterated over.

**Review by markstuart — COMMENTED** [review]
Sorry, the diff is a bit hard to parse, side by side diff probably makes a bit more sense when so much has changed.

File: ckanext/dia/commands.py [hunk 5]
```diff
@@ -69,22 +57,55 @@ def cleanup_datastore_once(self, start_offset):
...
+            end_chunk_index = offset + chunk_size
+            datastore_tables_chunk = all_datastore_tables[offset:end_chunk_index]
+            orphan_list = self._find_orphaned_datastore_tables(context, datastore_tables_chunk)
+            if len(orphan_list) > 0:
                 # Run a small chunk of the dataset to avoid locking up the
                 # database for a *really* long time(read: until postgres is
                 # restarted)
+                delete_count, error_list = self._delete_orphans(context, orphan_list)
+                total_deleted += delete_count
+                failed_deletes.extend(error_list)
```
Comment (tfentonz) [inline ckanext/dia/commands.py:76]: Is a test needed here to see if error_list has something in it? Or does extend not matter for empty lists?

File: ckanext/dia/commands.py (hunk 1 above)
Comment (tfentonz) [reply ckanext/dia/commands.py:100]: Is it worth having the ability to configure the limit?

**Review by tfentonz — APPROVED** [review]
✨ Looks super to me ✨

File: ckanext/dia/commands.py (hunk 1 above)
Comment (markstuart) [reply ckanext/dia/commands.py:100]: pass it in as argument to the script or something? I think the only reason you'd want to alter it would be to make it higher, and the timeout of 1 minute hardcoded into the datastore postgres adapter would likely be breached at that point anyway.

File: ckanext/dia/commands.py (hunk 5 above)
Comment (markstuart) [reply ckanext/dia/commands.py:76]: error list is initialised as an empty array in the `_delete_orphans` and yep, in python it's totally fine to `[].extend([])`
